cmd/lambda/lease_auth/auth.go

```diff
@@ -0,0 +1,185 @@
+package main
+
+import (
+	"encoding/json"
+	"fmt"
+	"github.com/Optum/dce/pkg/account/accountiface"
+	"github.com/Optum/dce/pkg/accountmanager/accountmanageriface"
+	"github.com/Optum/dce/pkg/api"
+	"github.com/Optum/dce/pkg/api/response"
+	"github.com/Optum/dce/pkg/errors"
+	leases "github.com/Optum/dce/pkg/lease"
+	"github.com/Optum/dce/pkg/lease/leaseiface"
+	"github.com/aws/aws-lambda-go/events"
+	"github.com/gorilla/mux"
+	"log"
+	"net/http"
+	"time"
+)
+
+func LeaseAuthHandler(w http.ResponseWriter, r *http.Request) {
+	// Retrieve the API Gateway context from our HTTP request object
+	// (required for our UserDetails service, to extract Cognito info)
+	reqCtx, err := muxLambda.GetAPIGatewayContext(r)
+	if err != nil {
+		log.Printf("Failed to parse context object from request: %s", err)
+		api.WriteAPIErrorResponse(w,
+			errors.NewInternalServer("Internal server error", err),
+		)
+		return
+	}
+
+	// Lease ID is optional.
+	// If not provided, we will attempt to find a matching
+	// lease, where PrincipalID == Cognito User ID
+	var leaseID *string
+	muxVars := mux.Vars(r)
+	if muxVars != nil {
+		leaseIDStr, ok := muxVars["leaseID"]
+		if ok {
+			leaseID = &leaseIDStr
+		}
+	}
+
+	// Generate credentials for the requested lease
+	res, err := leaseAuth(&leaseAuthInput{
+		leaseID:        leaseID,
+		requestContext: &reqCtx,
+		leaseService:   Services.LeaseService(),
+		accountService: Services.AccountService(),
+		userDetailer:   Services.UserDetailer(),
+		accountManager: Services.AccountManager(),
+		// Note that AWS has a hard 1-hour limit on roles assumed
+		// via "role-chaining" (assuming a role from an assumed-role).
+		// In our case, we're assuming a role, via our assumed lambda execution role.
+		principalSessionDuration: time.Hour,
+	})
+	if err != nil {
+		api.WriteAPIErrorResponse(w, err)
+		return
+	}
+
+	api.WriteAPIResponse(w, 201, res)
+}
+
+type leaseAuthInput struct {
+	leaseID                  *string
+	requestContext           *events.APIGatewayProxyRequestContext
+	leaseService             leaseiface.Servicer
+	accountService           accountiface.Servicer
+	userDetailer             api.UserDetailer
+	accountManager           accountmanageriface.Servicer
+	principalSessionDuration time.Duration
+}
+
+func leaseAuth(input *leaseAuthInput) (*response.LeaseAuthResponse, error) {
+	// Lookup the requesting user, via cognito
+	user := input.userDetailer.GetUser(input.requestContext)
+
+	var lease *leases.Lease
+	var err error
+	if input.leaseID != nil {
+		// Grab the lease from the DB
+		lease, err = input.leaseService.Get(*input.leaseID)
+		if err != nil {
+			return nil, err
+		}
+	} else if user.Username != "" {
```

Comment:
Is this a possible code path?  Either they have access to the lease or they don't?

Reply:
So logic here goes:

- if a Lease ID is provided, use that lease for auth'ing (line 81)
- If no lease is provided, but the request comes from an authenticated cognito user, lookup the active lease for that user (line 87)
- If we can't locate a lease using either of these approaches, return a 404 (line 101)